File: superflowsql/docker.py
"""Docker Compose operations - start, stop, status for the data stack."""

from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass
class ServiceStatus:
    """Status of a single Docker service."""

    name: str
    status: str  # running, stopped, not_found
    health: str  # healthy, unhealthy, starting, -
    ports: str

    @property
    def is_running(self) -> bool:
        return self.status == "running"

    @property
    def icon(self) -> str:
        if self.status == "running":
            return "●"
        elif self.status == "stopped":
            return "○"
        return "?"


def is_docker_running() -> bool:
    """Check if the Docker daemon is accessible."""
    try:
        subprocess.run(
            ["docker", "info"],
            capture_output=True,
            timeout=10,
        )
        return True
    except (subprocess.SubprocessError, FileNotFoundError):
        return False


def _docker_compose_cmd(project_dir: str | Path, *args: str) -> subprocess.CompletedProcess:
    """Run a docker compose command in the given project directory.

    Tries 'docker compose' (V2) first, falls back to 'docker-compose' (V1).
    """
    project_dir = str(Path(project_dir).resolve())

    # Try Docker Compose V2
    try:
        result = subprocess.run(
            ["docker", "compose", *args],
            cwd=project_dir,
            capture_output=True,
            text=True,
            timeout=300,
        )
        if result.returncode == 0:
            return result
    except FileNotFoundError:
        pass

    # Fallback to V1
    if shutil.which("docker-compose"):
        return subprocess.run(
            ["docker-compose", *args],
            cwd=project_dir,
            capture_output=True,
            text=True,
            timeout=300,
        )

    raise RuntimeError(
        "Docker Compose not found. Install Docker Desktop or docker-compose."
    )


def start_stack(project_dir: str | Path) -> str:
    """Start all services with docker compose up -d --build.

    Args:
        project_dir: Path to the SuperFlowSQL project directory.

    Returns:
        Combined stdout + stderr output from docker compose.
    """
    result = _docker_compose_cmd(project_dir, "up", "-d", "--build")
    output = (result.stdout + "\n" + result.stderr).strip()
    if result.returncode != 0:
        raise RuntimeError(f"Failed to start stack:\n{output}")
    return output


def stop_stack(project_dir: str | Path) -> str:
    """Stop all services with docker compose down.

    Args:
        project_dir: Path to the SuperFlowSQL project directory.

    Returns:
        Combined stdout + stderr output.
    """
    result = _docker_compose_cmd(project_dir, "down")
    output = (result.stdout + "\n" + result.stderr).strip()
    if result.returncode != 0:
        raise RuntimeError(f"Failed to stop stack:\n{output}")
    return output


def restart_stack(project_dir: str | Path) -> str:
    """Restart all services (down then up)."""
    stop_stack(project_dir)
    return start_stack(project_dir)


def get_stack_status(project_dir: str | Path) -> list[ServiceStatus]:
    """Get the status of all services in the stack.

    Returns:
        List of ServiceStatus for each running/stopped service.
    """
    try:
        result = _docker_compose_cmd(
            project_dir, "ps", "--format", "{{.Name}}|{{.Status}}|{{.Ports}}"
        )
    except RuntimeError:
        return _default_services()

    services: list[ServiceStatus] = []
    for line in result.stdout.strip().splitlines():
        line = line.strip()
        if not line:
            continue

        parts = line.split("|", 2)
        name = parts[0] if len(parts) >= 1 else "unknown"
        raw_status = parts[1].lower() if len(parts) >= 2 else ""
        ports = parts[2] if len(parts) >= 3 else ""

        if "up" in raw_status:
            status = "running"
            if "healthy" in raw_status:
                health = "healthy"
            elif "unhealthy" in raw_status:
                health = "unhealthy"
            else:
                health = "starting"
        else:
            status = "stopped"
            health = "-"

        services.append(ServiceStatus(name=name, status=status, health=health, ports=ports))

    return services if services else _default_services()


def _default_services() -> list[ServiceStatus]:
    """Return default expected services as not found."""
    return [
        ServiceStatus(name="postgres", status="not_found", health="-", ports=""),
        ServiceStatus(name="pgadmin", status="not_found", health="-", ports=""),
        ServiceStatus(name="airflow-webserver", status="not_found", health="-", ports=""),
        ServiceStatus(name="airflow-scheduler", status="not_found", health="-", ports=""),
        ServiceStatus(name="superset", status="not_found", health="-", ports=""),
    ]


def get_service_urls(config: dict | None = None) -> dict[str, str]:
    """Return default service access URLs.

    Args:
        config: Optional dict with port overrides (airflow_port, pgadmin_port, etc.)

    Returns:
        Dict mapping service names to their URLs.
    """
    cfg = config or {}
    return {
        "Airflow": f"http://localhost:{cfg.get('airflow_port', '8080')}",
        "PgAdmin": f"http://localhost:{cfg.get('pgadmin_port', '5050')}",
        "Superset": f"http://localhost:{cfg.get('superset_port', '8088')}",
        "PostgreSQL": f"localhost:{cfg.get('postgres_port', '5432')}",
    }

File: src/tui/core/docker.go
package core

import (
	"bytes"
	"fmt"
	"os/exec"
	"runtime"
	"strings"
)

// ServiceStatus represents the status of a Docker service
type ServiceStatus struct {
	Name   string
	Status string // "running", "stopped", "error", "not found"
	Port   string
	Health string
}

// StackAction represents an action to perform on the stack
type StackAction int

const (
	StackStart StackAction = iota
	StackStop
	StackRestart
)

// DockerComposeCmd runs a docker compose command in the project directory
func DockerComposeCmd(projectDir string, args ...string) (string, error) {
	// Try "docker compose" first (Docker Compose V2), fall back to "docker-compose"
	cmdArgs := append([]string{"compose"}, args...)
	cmd := exec.Command("docker", cmdArgs...)
	cmd.Dir = projectDir

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		// Try legacy docker-compose
		cmd2 := exec.Command("docker-compose", args...)
		cmd2.Dir = projectDir
		var stdout2, stderr2 bytes.Buffer
		cmd2.Stdout = &stdout2
		cmd2.Stderr = &stderr2

		err2 := cmd2.Run()
		if err2 != nil {
			return "", fmt.Errorf("docker compose failed: %s\n%s", stderr.String(), stderr2.String())
		}
		return stdout2.String(), nil
	}
	return stdout.String(), nil
}

// StartStack starts all services using docker compose
func StartStack(projectDir string) (string, error) {
	return DockerComposeCmd(projectDir, "up", "-d", "--build")
}

// StopStack stops all services
func StopStack(projectDir string) (string, error) {
	return DockerComposeCmd(projectDir, "down")
}

// GetStackStatus returns the status of all services
func GetStackStatus(projectDir string) ([]ServiceStatus, error) {
	output, err := DockerComposeCmd(projectDir, "ps", "--format", "{{.Name}}|{{.Status}}|{{.Ports}}")
	if err != nil {
		return nil, err
	}

	var services []ServiceStatus
	lines := strings.Split(strings.TrimSpace(output), "\n")

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.SplitN(line, "|", 3)
		s := ServiceStatus{
			Name:   "unknown",
			Status: "unknown",
			Port:   "",
		}
		if len(parts) >= 1 {
			s.Name = parts[0]
		}
		if len(parts) >= 2 {
			status := strings.ToLower(parts[1])
			if strings.Contains(status, "up") {
				s.Status = "running"
				if strings.Contains(status, "healthy") {
					s.Health = "healthy"
				} else if strings.Contains(status, "unhealthy") {
					s.Health = "unhealthy"
				} else {
					s.Health = "starting"
				}
			} else {
				s.Status = "stopped"
				s.Health = "-"
			}
		}
		if len(parts) >= 3 {
			s.Port = parts[2]
		}

		services = append(services, s)
	}

	// If no services found, return default expected services as "not found"
	if len(services) == 0 {
		services = []ServiceStatus{
			{Name: "postgres", Status: "not found", Health: "-"},
			{Name: "pgadmin", Status: "not found", Health: "-"},
			{Name: "airflow-webserver", Status: "not found", Health: "-"},
			{Name: "airflow-scheduler", Status: "not found", Health: "-"},
			{Name: "superset", Status: "not found", Health: "-"},
		}
	}

	return services, nil
}

// IsDockerRunning checks if Docker daemon is running
func IsDockerRunning() bool {
	cmd := exec.Command("docker", "info")
	return cmd.Run() == nil
}

// OpenBrowser opens a URL in the default browser
func OpenBrowser(url string) error {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "windows":
		cmd = exec.Command("cmd", "/c", "start", url)
	case "darwin":
		cmd = exec.Command("open", url)
	default:
		cmd = exec.Command("xdg-open", url)
	}
	return cmd.Run()
}
